🎉 Source Recurly: Support `incremental` Sync Mode

mohamagdy opened the pull request:
## What
Support `incremental` sync mode for Recurly

## How
After the successful migration of the Recurly source in Airbyte from using the `base_python` to CDK, changes the Recurly source to use the `AbstractSource` and `Stream` to implement the `incremental` sync mode.

## Recommended reading order
1. `source.py`
2. `streams.py`
3. Other files 

## 🚨 User Impact 🚨
Users had Recurly source already configured might require resetting the data after clicking the `Update latest source schema` from the connector page.


#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/SUMMARY.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

mohamagdy committed:
[1384] Change Recurly connector to use `AbstractSource` and `Stream`

Those 2 classes are part of the CDK and are the base line to start integrating
the `incremental` sync mode to the Recurly resources

mohamagdy committed:
[1384] Support `incremental` sync mode for Recurly

mohamagdy committed:
[1384] Add `account_coupon_redemptions` Recurly resource

mohamagdy committed:
[1384] Add Recurly source unit tests

mohamagdy committed:
Merge branch 'master' into 1384_recurly_incremental_sync_mode

mohamagdy committed:
Merge branch 'master' into 1384_recurly_incremental_sync_mode

mohamagdy committed:
Merge branch '1384_recurly_incremental_sync_mode' of github.com:mohamagdy/airbyte into 1384_recurly_incremental_sync_mode

mohamagdy committed:
[1384] Bump `source-recurly` version to `0.3.0`

Bumped the minor version as the `incremental` sync mode functionality is added

mohamagdy committed:
[1384] Removed `.python-version` from Git

mohamagdy committed:
[1384] Annotate the Recurly `api_key` with `airbyte_secret`

mohamagdy commented:
/test connector=source-recurly

mohamagdy commented:
```shell

> Task :airbyte-integrations:connectors:source-recurly:unitTest
[python] .venv/bin/python -m pytest -s unit_tests --cov=./ --cov-config=/Users/mohamedmagdy/projects/canary/airbyte/tools/python/.coveragerc --cov-report html --cov-report term
         ============================= test session starts ==============================
         platform darwin -- Python 3.9.7, pytest-6.1.2, py-1.11.0, pluggy-0.13.1 -- /Users/mohamedmagdy/projects/canary/airbyte/airbyte-integrations/connectors/source-recurly/.venv/bin/python
         cachedir: .pytest_cache
         rootdir: /Users/mohamedmagdy/projects/canary/airbyte, configfile: pytest.ini
         plugins: cov-2.12.1, sugar-0.9.4, mock-3.6.1, timeout-1.4.2
         collecting ... collected 6 items

         unit_tests/test_streams.py::TestStreams::test_account_coupon_redemptions_read_records PASSED
         unit_tests/test_streams.py::TestStreams::test_export_dates_read_records PASSED
         unit_tests/test_streams.py::TestStreams::test_get_updated_state PASSED
         unit_tests/test_streams.py::TestStreams::test_measured_unit_client_method_name PASSED
         unit_tests/test_streams.py::TestStreams::test_read_records PASSED
         unit_tests/test_streams.py::TestStreams::test_read_records_with_begin_time PASSED

         ---------- coverage: platform darwin, python 3.9.7-final-0 -----------
         Name                         Stmts   Miss  Cover
         ------------------------------------------------
         source_recurly/__init__.py       2      0   100%
         source_recurly/source.py        24     13    46%
         source_recurly/streams.py       82      7    91%
         ------------------------------------------------
         TOTAL                          108     20    81%
         Coverage HTML written to dir htmlcov

```

mohamagdy commented:
```shell
         ---------- coverage: platform darwin, python 3.7.9-final-0 -----------
         Name                                                 Stmts   Miss  Cover
         ------------------------------------------------------------------------
         source_acceptance_test/__init__.py                       2      0   100%
         source_acceptance_test/base.py                           8      4    50%
         source_acceptance_test/config.py                        74      8    89%
         source_acceptance_test/conftest.py                     109    109     0%
         source_acceptance_test/plugin.py                        47     47     0%
         source_acceptance_test/tests/__init__.py                 4      0   100%
         source_acceptance_test/tests/test_core.py              235     95    60%
         source_acceptance_test/tests/test_full_refresh.py       38     27    29%
         source_acceptance_test/tests/test_incremental.py        69     38    45%
         source_acceptance_test/utils/__init__.py                 6      0   100%
         source_acceptance_test/utils/asserts.py                 37      2    95%
         source_acceptance_test/utils/common.py                  54     24    56%
         source_acceptance_test/utils/compare.py                 60     25    58%
         source_acceptance_test/utils/connector_runner.py        80     49    39%
         source_acceptance_test/utils/json_schema_helper.py     115     14    88%
         ------------------------------------------------------------------------
         TOTAL                                                  938    442    53%
         Coverage HTML written to dir htmlcov


         Results (9.06s):
              107 passed
```

mohamagdy committed:
[1384] Bump `source-recurly` to `0.3.0`

mohamagdy committed:
[1384] Update code style

After running `gradle`

mohamagdy committed:
[1384] Update Recurly source documentation

By adding the Account Coupon Redemption documentation
and setting the `Incremental sync` to `Yes`

gerhalt commented on airbyte-integrations/connectors/source-recurly/source_recurly/source.py:
```diff
@@ -2,71 +2,63 @@
+
+        self.__client = None
+
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, Optional[Any]]:
```
This class looks completely different, I'm assuming that's due to the significant framework changes since the original was written?

mohamagdy replied:
Yes, the Recurly source was rewritten using CDK in a previous PR (it was using a legacy `base_python` code). This time I did another round of refactoring and inherited from some CDK base classes and implemented the abstract methods.

gerhalt commented on airbyte-integrations/connectors/source-recurly/source_recurly/streams.py:
```diff
@@ -0,0 +1,219 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
```
Is it? You've run their license by legal?

mohamagdy replied:
This is autogenerated when I ran the `gradle` build tool that runs some acceptance, integration and unit tests as well as code styling 😕

gerhalt commented on airbyte-integrations/connectors/source-recurly/source_recurly/streams.py:
```diff
@@ -0,0 +1,219 @@
+    @abstractmethod
+    def name(self):
+        """
+        The name of the Recurly resource. All children inherits from this class should set the resource name
```
grammar

mohamagdy replied:
Fixed in https://github.com/mohamagdy/airbyte/pull/1/commits/54786be7856f1fb79dfe0306785fb3e18f00cef7

gerhalt commented on airbyte-integrations/connectors/source-recurly/source_recurly/streams.py:
```diff
@@ -0,0 +1,219 @@
+        resources have different primary key or a composite key can override this method.
+
+        :return: The Recurly resource primary key(s)
+        :rtype: Either `str`, list(str) or list(list(str))
```
:(

mohamagdy replied:
Inherited https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/core.py#L106

gerhalt commented on airbyte-integrations/connectors/source-recurly/source_recurly/streams.py:
```diff
@@ -0,0 +1,219 @@
+
+        self.begin_time = (stream_state and stream_state[self.cursor_field]) or self.begin_time
+
+        if self.begin_time:
+            params.update({self.BEGIN_TIME_PARAM: self.begin_time})
```
This limits to only newer records? Are there any other filters needed here?

gerhalt commented on airbyte-integrations/connectors/source-recurly/source_recurly/streams.py:
```diff
@@ -0,0 +1,219 @@
+            return resource
+
+
+class RecurlyAccountsStream(BaseRecurlyStream):
```
I like these simple subclasses.

gerhalt commented on airbyte-integrations/connectors/source-recurly/source_recurly/streams.py:
```diff
@@ -0,0 +1,219 @@
+    ) -> Iterable[Mapping[str, Any]]:
+        """
+        The method to be called to retrieve the accounts coupon redemptions from Recurly. To retrieve the coupon
+        redemptions, a seprate call to list all the accounts should be made to pass the `account_id` to the account
```
sp

mohamagdy replied:
Fixed the typo in https://github.com/mohamagdy/airbyte/pull/1/commits/d2510636013af58d464fe9b170f3bde0360b49dc

gerhalt commented on airbyte-integrations/connectors/source-recurly/source_recurly/streams.py:
```diff
@@ -0,0 +1,219 @@
+        """
+        params = {"order": "asc", "sort": self.sort_key}
+
+        self.begin_time = (stream_state and stream_state[self.cursor_field]) or self.begin_time
```
Stream state is only populated if we're already streaming records?

mohamagdy replied:
Correct. The `stream_state` is set in case of `incremetal` sync mode to keep track of the cursor value that is used to fetch resources changed after that cursor value. `get_updated_state` method implements the logic of setting the stream state.

mohamagdy committed:
[1384] Fix typo in Python doc

mohamagdy committed:
[1384] List all accounts when fetching the redeemed coupons

As fetching accounts only updated after the `begin_time` won't work in
`incremental` sync mode if an older account redeemed coupons

mohamagdy committed:
[1384] Fix typo in Python doc

mohamagdy commented:
Closing in favor of https://github.com/airbytehq/airbyte/pull/8675
